### PR TITLE
algod: Txn Group Delta Apis

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary

Implements two new APIs:
* /v2/deltas/{round}/txn/group
* /v2/deltas/txn/group/{id}

These expose the state delta objects for groups of transactions.

## Test Plan

Adding handler tests as the PR progresses.

### CODE REVIEW

File: daemon/algod/api/server/v2/handlers.go [hunk 1]
```diff
@@ -91,6 +92,7 @@ type LedgerForAPI interface {
 	Block(rnd basics.Round) (blk bookkeeping.Block, err error)
 	AddressTxns(id basics.Address, r basics.Round) ([]transactions.SignedTxnWithAD, error)
 	GetStateDeltaForRound(rnd basics.Round) (ledgercore.StateDelta, error)
+	GetTracer() logic.EvalTracer
```
Comment (winder) [inline daemon/algod/api/server/v2/handlers.go:95]: nit: Should this return an interface that gets cast to the EvalTracer?

File: daemon/algod/api/server/v2/handlers.go [hunk 2]
```diff
@@ -1703,6 +1705,54 @@ func (v2 *Handlers) TealDisassemble(ctx echo.Context) error {
...
+		return badRequest(ctx, err, errFailedParsingFormatOption, v2.Log)
+	}
+	idDigest, err := crypto.DigestFromString(id)
+	if err != nil {
+		return badRequest(ctx, err, errNoValidTxnSpecified, v2.Log)
+	}
+	tracer, ok := v2.Node.LedgerForAPI().GetTracer().(*eval.TxnGroupDeltaTracer)
+	if !ok {
+		return serviceUnavailable(ctx, err, errFailedRetrievingTracer, v2.Log)
+	}
```
Comment (winder) [inline daemon/algod/api/server/v2/handlers.go:1722]: Does this handle a nil tracer?

File: ledger/eval/txntracer.go [hunk 3]
```diff
@@ -72,29 +102,34 @@ func (tracer *TxnGroupDeltaTracer) AfterTxnGroup(ep *logic.EvalParams, deltas *l
 }
 
 // GetDeltasForRound supplies all StateDelta objects for txn groups in a given rnd
-func (tracer *TxnGroupDeltaTracer) GetDeltasForRound(rnd basics.Round) ([]ledgercore.StateDelta, error) {
+func (tracer *TxnGroupDeltaTracer) GetDeltasForRound(rnd basics.Round) ([]TxnGroupDeltaWithIds, error) {
 	rndEntries, exists := tracer.txnGroupDeltas[rnd]
```
Comment (winder) [inline ledger/eval/txntracer.go:106]: It looks like `txnGroupDeltas` needs a RW lock.

File: daemon/algod/api/algod.oas2.json [hunk 4]
```diff
@@ -1534,6 +1534,146 @@
...
+          },
+          "default": {
+            "description": "Unknown Error"
+          }
+        }
+      }
+    },
+    "/v2/deltas/txn/group/{id}": {
+      "get": {
+        "description": "Get ledger deltas for a given transaction group.",
```
Comment (algochoi) [inline daemon/algod/api/algod.oas2.json:1609]: Correct me if I'm wrong - it sounds like there is one ledger delta returned for the entire group:
```suggestion
        "description": "Get a ledger delta for a given transaction group.",
```

File: daemon/algod/api/algod.oas2.json [hunk 5]
```diff
@@ -2520,6 +2660,26 @@
       "type": "object",
       "x-algorand-format": "StateDelta"
     },
+    "LedgerStateDeltaForTransactionGroup": {
+      "description": "Contains ledger deltas for a single transaction group",
```
Comment (algochoi) [inline daemon/algod/api/algod.oas2.json:2664]:
```suggestion
      "description": "Contains a ledger delta for a single transaction group",
```

**Review by algochoi — COMMENTED** [review]
I created a child PR while going through this PR to add some additional handler checks for my own sanity: https://github.com/Eric-Warehime/go-algorand/pull/7

File: daemon/algod/api/server/v2/handlers.go (hunk 2 above)
Comment (Eric-Warehime) [reply daemon/algod/api/server/v2/handlers.go:1722]: Yes, I've added a test for this case as well just to be sure.

I'm not sure that 503 is the exactly the correct http response though--open to suggestions on better responses there.

**Review by algochoi — APPROVED** [review]

File: daemon/algod/api/server/v2/handlers.go (hunk 2 above)
Comment (winder) [reply daemon/algod/api/server/v2/handlers.go:1722]: I don't think 503 is a good choice, isn't it supposed to be fore temporary conditions? I'd probably go with 501 or 500.

File: daemon/algod/api/algod.oas2.json [hunk 6]
```diff
@@ -1534,6 +1534,146 @@
...
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          },
+          "500": {
+            "description": "Internal Error",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          },
+          "503": {
```
Comment (algochoi) [inline daemon/algod/api/algod.oas2.json:1595]: I think this became a 501 in the recent commit

File: daemon/algod/api/algod.oas2.json [hunk 7]
```diff
@@ -1534,6 +1534,146 @@
...
```
Comment (algochoi) [inline daemon/algod/api/algod.oas2.json:1665]: 501 - not implemented

**Review by algochoi — COMMENTED** [review]
Just need to regenerate the specs if we change the 503 to 501?

**Review by algochoi — DISMISSED** [review]
LGTM, I am not particularly opinionated about what error should be returned for a nil tracer.

File: daemon/algod/api/algod.oas2.json [hunk 8]
```diff
@@ -2520,6 +2660,26 @@
...
+      ],
+      "properties": {
+        "delta": {
+          "$ref": "#/definitions/LedgerStateDelta"
+        },
+        "ids": {
+          "type": "array",
+          "items": {
+            "type": "string",
+            "x-algorand-format": "Address"
```
[the remark this answers is not in the document]
Comment (winder) [reply daemon/algod/api/algod.oas2.json:2678]: When would there be multiple IDs for a state delta? This is each txid in the group?

File: daemon/algod/api/server/v2/handlers.go [hunk 9]
```diff
@@ -22,6 +22,7 @@ import (
 	"encoding/base64"
 	"errors"
 	"fmt"
+	"github.com/algorand/go-algorand/ledger/eval"
```
Comment (winder) [inline daemon/algod/api/server/v2/handlers.go:25]: nit: import groups

**Review by winder — CHANGES_REQUESTED** [review]
Just one small change needed in the OpenAPI, otherwise LGTM.

File: daemon/algod/api/algod.oas2.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (Eric-Warehime) [reply daemon/algod/api/algod.oas2.json]: The Ids for a given delta is the set of all Txn IDs and/or Group IDs associated with that delta.

**Review by winder — APPROVED** [review]

**Review by algochoi — APPROVED** [review]

**Pull request merged**